### ISSUE TITLE
How to add Apptainer Support for Intel GPUs

### ISSUE THREAD

**Issue by sramakintel** [apptainer/apptainer#1592]
Apptainer supports NVIDIA and AMD GPUs-https://apptainer.org/docs/user/main/gpu.html.  How to add support for Intel discrete GPUs? (https://www.intel.com/content/www/us/en/products/docs/discrete-gpus/data-center-gpu/max-series-gpu-product-brief.html#:~:text=Intel%20Data%20Center%20GPU%20Max%20Series%20%20,L2%20108%20MB%20%204%20more%20rows%20) What is the process to contribute?

**Comment by upsj** [comment]
https://github.com/apptainer/apptainer/issues/841 has some details

**Comment by upsj** [comment]
I was just playing around with `--rocm` binds and noticed that since `/dev/dri` is being mounted completely, `--rocm` already does half of what is necessary to run Intel GPUs. The other half is either installing the necessary libraries inside the container or (better) forwarding them from outside (like with ROCm)

The libraries I installed are
```
libiga64.so.1.0.13822.6
libigc.so.1.0.13822.6
libiga64.so
libiga64.so.1
libigc.so
libigc.so.1
libigdfcl.so.1.0.13822.6
libopencl-clang.so.11
libigdfcl.so
libigdfcl.so.1
libopencl-clang.so
libze_intel_gpu.so.1
intel-opencl/libigdrcl.so
libocloc.so
libigdgmm.so.12
```
but as you seem affiliated with Intel, you probably know best what's necessary :)

So by copying the ROCm binding approach and creating a custom `intelliblist.conf` similar to `rocmliblist.conf` with the aforementioined libraries and maybe `xpu-smi` and `sycl-ls`, you should be able to do this easily.

Full execution, running on [ginkgodev/unified:latest](https://hub.docker.com/r/ginkgodev/unified/tags)
```bash
$ apptainer shell --rocm --writable-tmpfs unified_latest.sif 
Apptainer> /usr/bin/entrypoint.sh bash # load the necessary spack modules etc
$ sycl-ls
[opencl:gpu:2] Intel(R) OpenCL Graphics, Intel(R) Arc(TM) A770 Graphics 3.0 [23.17.26241.22]
$ cd oneAPI-samples/DirectProgramming/C++SYCL/DenseLinearAlgebra/simple-add && mkdir build && cd build && CXX=dpcpp cmake .. && ninja
...
$ ONEAPI_DEVICE_SELECTOR=*:gpu ./simple-add-buffers 
Running on device: Intel(R) Arc(TM) A770 Graphics
Array size: 10000
[0]: 0 + 100000 = 100000
[1]: 1 + 100000 = 100001
[2]: 2 + 100000 = 100002
...
[9999]: 9999 + 100000 = 109999
Successfully completed on device.

**Comment by sramakintel** [comment]
> I was just playing around with `--rocm` binds and noticed that since `/dev/dri` is being mounted completely, `--rocm` already does half of what is necessary to run Intel GPUs. The other half is either installing the necessary libraries inside the container or (better) forwarding them from outside (like with ROCm)
> 
> The libraries I installed are
> 
> ```
> libiga64.so.1.0.13822.6
> libigc.so.1.0.13822.6
> libiga64.so
> libiga64.so.1
> libigc.so
> libigc.so.1
> libigdfcl.so.1.0.13822.6
> libopencl-clang.so.11
> libigdfcl.so
> libigdfcl.so.1
> libopencl-clang.so
> libze_intel_gpu.so.1
> intel-opencl/libigdrcl.so
> libocloc.so
> libigdgmm.so.12
> ```
> 
> but as you seem affiliated with Intel, you probably know best what's necessary :)
> 
> So by copying the ROCm binding approach and creating a custom `intelliblist.conf` similar to `rocmliblist.conf` with the aforementioined libraries and maybe `xpu-smi` and `sycl-ls`, you should be able to do this easily.
> 
> Full execution, running on [ginkgodev/unified:latest](https://hub.docker.com/r/ginkgodev/unified/tags)
> 
> ```shell
> $ apptainer shell --rocm --writable-tmpfs unified_latest.sif 
> Apptainer> /usr/bin/entrypoint.sh bash # load the necessary spack modules etc
> $ sycl-ls
> [opencl:gpu:2] Intel(R) OpenCL Graphics, Intel(R) Arc(TM) A770 Graphics 3.0 [23.17.26241.22]
> $ cd oneAPI-samples/DirectProgramming/C++SYCL/DenseLinearAlgebra/simple-add && mkdir build && cd build && CXX=dpcpp cmake .. && ninja
> ...
> $ ONEAPI_DEVICE_SELECTOR=*:gpu ./simple-add-buffers 
> Running on device: Intel(R) Arc(TM) A770 Graphics
> Array size: 10000
> [0]: 0 + 100000 = 100000
> [1]: 1 + 100000 = 100001
> [2]: 2 + 100000 = 100002
> ...
> [9999]: 9999 + 100000 = 109999
> Successfully completed on device.
> ```

This looks great. Thanks for the inputs. Quick question, using `--rocm` would load the `rocmliblist.conf`. How does the `intelliblist.conf` get called when `--rocm` is used?

**Comment by upsj** [comment]
It isn't, I was talking about a hypothetical `--intel` flag and how it could be implemented (starting from the ROCm code seems like a good approach)

**Comment by pradt2** [comment]
Hi, is there an update on this? If not, I can try to contribute support for this but I will need project guidance as I've not contributed to Apptainer before?

**Comment by upsj** [comment]
I would argue that apptainer already has Intel GPU support (I'm running CI in production using apptainer on multiple Intel GPU nodes) via the `--rocm` flag, albeit not documented, since that flag mounts the `/dev/dri` and `/dev/kfd` devices.
All you need is a container with the necessary runtime libraries available (like the [Intel oneAPI Basekit](https://hub.docker.com/r/intel/oneapi-basekit) Docker containers).
There has been a bit of a discussion about which libraries should be mounted from the host side to facilitate GPU support (if any), in my experience there are no libraries that need to be provided by the host for Intel GPUs to work properly for SYCL via the OpenCL or Level Zero backends.

**Comment by sramakintel** [comment]
> I would argue that apptainer already has Intel GPU support (I'm running CI in production using apptainer on multiple Intel GPU nodes) via the `--rocm` flag, albeit not documented, since that flag mounts the `/dev/dri` and `/dev/kfd` devices. All you need is a container with the necessary runtime libraries available (like the [Intel oneAPI Basekit](https://hub.docker.com/r/intel/oneapi-basekit) Docker containers). There has been a bit of a discussion about which libraries should be mounted from the host side to facilitate GPU support (if any), in my experience there are no libraries that need to be provided by the host for Intel GPUs to work properly for SYCL via the OpenCL or Level Zero backends.

Yes this is corrected. I have verified too. The docker container like `intel/intel-extension-for-pytorch:2.3.110-xpu` has the OneAPI runtime libraries and User mode GPU drivers and PyTorch that are needed to detect the GPU and offload compute onto them. No special upstream is required for Intel GPUs

**Comment by pradt2** [comment]
That's good news, but not the end of the story. One of the reasons why we have the gpu flags is so that our images don't have to contain all the drivers/low level system libs for GPU support. At least the case with Nvidia is that low level libs need to match the host driver, and finding out the correct versions, and how to get them on a particular image distro is difficult. I'd imagine that Intel is similar.

It's good to hear that `intel/intel-extension-for-pytorch:2.3.110-xpu` works with `--rocm` but it's not a universally applicable solution. I can't use this image as a base because I already have an existing image based on a specific distro, and migrating it over to this would take too much time. Also, more generally, apptainer support for Intel GPUs should not rely on an external Docker image that's not maintained by Apptainer.

To keep things consistent across the gpu vendors, I really think there should be a dedicated flag for Intel, and the low level libs should be exposed to the image.

**Comment by pradt2** [comment]
Thanks for shedding some light onto this. Now I understand the hesitation to add a dedicated .conf file.

What's you stance on adding a dedicated flag similar to --nv or --rocm for Intel though? I think it still makes sense even if users can achieve the same effect using the --rocm flag, because right now using --rocm is an undocumented **workaround**, and adding a dedicated flag would indicate at least some level of **support** for the hardware.

**Comment by pradt2** [comment]
@upsj You mentioned an internal Slack, maybe it would make sense to move the discussion there?

**Comment by upsj** [comment]
We don't have a premium account IIRC, so discussions there disappear after a while. I think it would be more permanent to put it here.

**Comment by DrDaveD** [comment]
Fixed by @upsj in #2606
